lib/tools/system-calls.js

```diff
@@ -558,7 +559,13 @@ systemCallMethods.reboot = async function (retries = DEFAULT_ADB_REBOOT_RETRIES)
 
 systemCallMethods.root = async function () {
   try {
-    await exec(this.executable.path, ['root']);
+    let {stdout} = await exec(this.executable.path, ['root']);
+
+    // on real devices in some situations we get an error in the stdout
+    if (stdout.indexOf('adbd cannot run as root') !== -1) {
```

Comment:
It is `adb` that is not returning a non-zero code. All `exec` does is look at what is execed. It cannot, _a priori_, know that any particular output indicates a failure.